packages/beacon-node/src/network/core/workerNetworkCore.ts

```diff
@@ -0,0 +1,188 @@
+import {PublishResult} from "@libp2p/interface-pubsub";
+import {exportToProtobuf} from "@libp2p/peer-id-factory";
+import {PeerId} from "@libp2p/interface-peer-id";
+import {Multiaddr} from "@multiformats/multiaddr";
+import {routes} from "@lodestar/api";
+import {PeerScoreStatsDump} from "@chainsafe/libp2p-gossipsub/dist/src/score/peer-score.js";
+import {allForks, altair, deneb, phase0} from "@lodestar/types";
+import {PublishOpts} from "@chainsafe/libp2p-gossipsub/types";
+import {spawn, Thread, Worker} from "@chainsafe/threads";
+import {BeaconConfig, chainConfigToJson} from "@lodestar/config";
+import {NetworkEvent, NetworkEventBus} from "../events.js";
+import {CommitteeSubscription} from "../subnets/interface.js";
+import {PeerScoreStats} from "../peers/index.js";
+import {NetworkOptions} from "../options.js";
+import {NetworkWorkerApi, NetworkWorkerData, NetworkCore} from "./types.js";
+
+export type Libp2pkWorkerWrapperInitModules = {
+  opts: NetworkOptions;
+  config: BeaconConfig;
+  genesisTime: number;
+  activeValidatorCount: number;
+  peerId: PeerId;
+  events: NetworkEventBus;
+};
+
+type Libp2pkWorkerWrapperModules = Libp2pkWorkerWrapperInitModules & {
+  workerApi: NetworkWorkerApi;
+};
+
+/**
+ * NetworkCore implementation using a Worker thread
+ */
+export class WorkerNetworkCore implements NetworkCore {
+  constructor(private readonly modules: Libp2pkWorkerWrapperModules) {}
+
+  static async init(modules: Libp2pkWorkerWrapperInitModules): Promise<WorkerNetworkCore> {
+    const {opts, config, genesisTime, peerId, events} = modules;
+
+    const workerData: NetworkWorkerData = {
+      opts,
+      chainConfigJson: chainConfigToJson(config),
+      genesisValidatorsRoot: config.genesisValidatorsRoot,
+      peerIdProto: exportToProtobuf(peerId),
+      bindAddr: opts.bindAddr,
+      metrics: opts.metrics,
+      peerStoreDir: opts.peerStoreDir,
+      genesisTime,
+      activeValidatorCount: modules.activeValidatorCount,
+    };
+
+    const worker = new Worker("./worker.js", {workerData} as ConstructorParameters<typeof Worker>[1]);
+
+    const workerApi = await spawn<NetworkWorkerApi>(worker, {
+      // A Lodestar Node may do very expensive task at start blocking the event loop and causing
+      // the initialization to timeout. The number below is big enough to almost disable the timeout
+      timeout: 5 * 60 * 1000,
+    });
+
+    workerApi.pendingGossipsubMessage().subscribe((data) => events.emit(NetworkEvent.pendingGossipsubMessage, data));
```

Comment:
Difference to discv5, this worker will have much more events. Having to do this observable pattern for each sounds unnecessary. We could emit all network events on a single observable, and then emit to their respective topics on the other side